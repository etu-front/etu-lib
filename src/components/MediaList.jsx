import PropTypes from 'prop-types'
import React, { PureComponent } from 'react'
import LazyLoad from 'react-lazyload'
import styled from 'styled-components'
import { Card, Icon, Tooltip } from 'antd'
import FileIcon from './FileIcon'

import dnd from './dnd'

const PlayIcon = styled(Icon)`
  color: white;
  font-size: 28px;
  align-self: center;
  margin: 0 auto;
  &:hover {
    color: #eee;
  }
`

const ImageCard = ({ thumbnail, width, height, backgroundColor = '#F0F0F0', icon }) => (
  <div
    style={{
      backgroundImage: `url(${thumbnail})`,
      backgroundSize: 'cover',
      backgroundPosition: 'center',
      display: 'flex',
      width,
      height,
      margin: '0 auto',
      maxWidth: '100%',
      backgroundColor
    }}
  >
    {icon}
  </div>
)

export const List = styled.div`
  clear: both;
  margin: 0 -${props => props.margin}px;
  list-style: none;
  display: flex;
  flex-wrap: wrap;
  flex-direction: row;
  & > div, > .ant-card {
    margin: ${props => props.margin + 'px'};
    overflow: hidden;
    border-color: #d5d4d4;
    border-radius: 4px;
    background: ${props => props.background};
    position: relative;
    display: flex;
    align-items: ${props => (props.fixed ? 'center' : 'flex-start')};
    justify-content: center;
  }
`
List.Item = styled(Card)`
  &.ant-card {
    margin: ${props => props.margin + 'px'};
    overflow: hidden;
    border-color: #d5d4d4;
    border-radius: 4px;
    background: ${props => props.background};
    position: relative;
    display: flex;
    align-items: ${props => (props.fixed ? 'center' : 'flex-start')};
    justify-content: center;
  }
  .ant-card-body {
    padding: ${props => (props.fixed ? 0 : '18px 22px')};
    text-align: center;
    cursor: pointer;
  }
`
List.Item.displayName = 'ListItem'

List.Name = styled.div`
  position: absolute;
  left: 0;
  right: 0;
  padding: 0 5px;
  bottom: ${props => (props.fixed ? 0 : '10px')};
  height: 20px;
  font-size: 12px;
  line-height: 20px;
  color: ${props => (props.fixed ? '#fff' : '#787878')};
  background: ${props => (props.fixed ? '#33333399' : 'transparent')};
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  text-align: center;
`

export const FileName = ({ name, fixed }) => (
  <List.Name fixed={fixed ? 1 : 0}>
    <Tooltip title={name.length < 20 ? '' : name} placement="topLeft">
      {name}
    </Tooltip>
  </List.Name>
)

export const Close = styled.div`
  position: absolute;
  right: 2px;
  top: 2px;
  color: red;
  cursor: pointer;
  .anticon-iconfont {
    font-size: 18px;
  }
`
class MediaList extends PureComponent {
  static displayName = 'MediaList'
  static propTypes = {
    items: PropTypes.array.isRequired,
    noName: PropTypes.bool,
    margin: PropTypes.number,
    bordered: PropTypes.bool,
    fixed: PropTypes.bool,
    onDelete: PropTypes.func,
    hasDeleteButton: PropTypes.bool,
    pdfPrevView: PropTypes.bool,
    lazyLoad: PropTypes.bool,
    onClick: PropTypes.func,
    onSort: PropTypes.func,
    sortOptions: PropTypes.shape({
      type: PropTypes.string,
      overClassName: PropTypes.string,
      draggingClassName: PropTypes.string,
      overStyle: PropTypes.object,
      draggingStyle: PropTypes.object
    })
  }

  static defaultProps = {
    items: [],
    background: '#f5f5f5',
    width: 100,
    height: 100,
    margin: 5,
    bordered: false,
    fixed: true,
    noName: true,
    pdfPrevView: false,
    lazyLoad: true,
    sortOptions: {
      type: 'SORT_ITEM',
      overClassName: '',
      draggingClassName: '',
      overStyle: {},
      draggingStyle: {}
    }
  }

  static getThumbnail = (id, width, height) => {}
  static getFile = id => {}
  static getDownload = id => {}
  // 绑定获取文件地址函数
  static bindGetFileFunctions = ({ getThumbnail, getFile, getDownload }) => {
    MediaList.getFile = getFile
    MediaList.getThumbnail = getThumbnail
    MediaList.getDownload = getDownload
  }

  static getFileName = data => {
    if (data.name) return data.name
    if (data.extra && data.extra.file_name) return data.extra.file_name
    if (data.extra && data.extra.attrs) return data.extra.attrs.file_name
  }

  static getFileId = item => {
    if (item.data) return item.data
    return item.media_id || item.file_id || item.id
  }

  // static getDerivedStateFromProps(props, state) {
  //   if (props.items !== state.items) {
  //     return {
  //       ...state,
  //       items: props.items
  //     }
  //   }
  //   return state
  // }
  // state = {
  //   items: this.props.items
  // }

  handleRemoveItem = (id, e) => {
    e.preventDefault()
    e.stopPropagation()
    this.props.onDelete && this.props.onDelete(id)
  }

  onClick = (items, index) => {
    const item = items[index]
    if (item.type !== 'picture' && item.type !== 'video') {
      const a = window.document.createElement('a')
      const fileId = MediaList.getFileId(item)
      a.href = this.props.pdfPrevView && item.type === 'pdf' ? MediaList.getFile(fileId) : MediaList.getDownload(fileId)
      a.target = '_blank'
      document.body.appendChild(a)
      a.click()
      setTimeout(function() {
        document.body.removeChild(a)
      }, 0)
    } else {
      this.props.onClick && this.props.onClick(items, index)
    }
  }

  sortItem = (dragIndex, hoverIndex) => {
    const items = [...this.props.items]
    const item = items[dragIndex]
    if (!item) return
    items.splice(dragIndex, 1)
    items.splice(hoverIndex, 0, item)
    this.props.onSort(items)
  }

  renderMediaNode = data => {
    const { width, height, lazyLoad, fixed } = this.props
    const fileId = MediaList.getFileId(data)
    const mediaType = data.type

    let thumbnail,
      imgWidth = width / (fixed ? 1 : 1.8),
      imgHeight = height / (fixed ? 1 : 2.5)

    thumbnail = MediaList.getThumbnail(fileId, width, height)
    // 预览模式，小图
    if (mediaType === 'video' || mediaType === 'picture') {
      const img = (
        <ImageCard
          thumbnail={thumbnail}
          width={imgWidth}
          height={imgHeight}
          icon={mediaType === 'video' && <PlayIcon type="play-circle" />}
        />
      )
      if (lazyLoad) {
        return (
          <LazyLoad height={height} offset={50} once>
            {img}
          </LazyLoad>
        )
      }
      return img
    }
    return <FileIcon type={mediaType} size={fixed ? width / 2 : imgHeight} style={{ margin: '0 auto' }} />
  }

  // 视频和图片
  render() {
    const {
      margin,
      bordered,
      background,
      noName,
      fixed,
      width,
      height,
      hasDeleteButton = false,
      onSort,
      sortOptions
    } = this.props
    const { items } = this.props
    const nodes = items.map((item, index) => {
      const fileName =
        (!noName || (item.type !== 'picture' && item.type !== 'video')) && MediaList.getFileName(item)
      const fileId = MediaList.getFileId(item)
      const key = fileId + '_' + index
      const node = (
        <List.Item
          onClick={this.onClick.bind(null, items, index)}
          key={key}
          bordered={bordered && !noName}
          style={{ width, height }}
          fixed={fixed ? 1 : 0}
          margin={margin}
          background={background}
        >
          {this.renderMediaNode(item)}
          {hasDeleteButton && (
            <Close onClick={this.handleRemoveItem.bind(null, fileId)}>
              <Icon type="delete" />
            </Close>
          )}
          {fileName && <FileName fixed={fixed} name={fileName} />}
        </List.Item>
      )
      if (nodes.length > 1 && onSort) {
        const { type: sortType, ...sortProps } = sortOptions
        const Sortable = dnd.generateSortable(sortType)
        return <Sortable key={key} index={index} onSort={this.sortItem} {...sortProps}>{node}</Sortable>
      }
      return node
    })
    return <List margin={margin}>{nodes}</List>
  }
}

export default dnd.wrap(MediaList)
