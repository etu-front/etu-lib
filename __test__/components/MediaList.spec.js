import React from 'react'
import MediaList, { List, FileName } from '../../src/components/MediaList'
import { shallow, render, mount } from 'enzyme'
import renderer from 'react-test-renderer'

export const getThumbnail = (id, width, height) =>
  `/bos/api/v1/files/${id}/thumbnail/?width=${width}&height=${height}&mode=zoom_cut`
export const getFile = id => `/bos/api/v1/files/${id}/`
export const getDownload = id => `/bos/api/v1/files/${id}/download/`

const setup = () => {
  MediaList.bindGetFileFunctions({ getThumbnail, getFile, getDownload })

  const medias = [
    {
      type: 'pdf',
      media_id: 'E77SKH7F7HXQ06YNZAF64GFC76Q9SEZA',
      id: 'E77SKH7F7HXQ06YNZAF64GFC76Q9SEZA',
      name: 'LastOfUs2.pdf'
    },
    {
      course_resource_id: 163,
      data: 'YM0JXBCAA84DJ7WEZ20TY5D4HZV3DHW3',
      id: 381,
      name: '【中文版】ETU PBL Backwards Design Template.docx',
      type: 'word',
      media_id: 'YM0JXBCAA84DJ7WEZ20TY5D4HZV3DHW3'
    },
    {
      media_id: '67BD020K2E6JMGX9R00NZ8DQ8KBSSWPG',
      type: 'picture',
      name: '1530871462174.jpg',
      extra: {
        url: '/api/v1/files/67BD020K2E6JMGX9R00NZ8DQ8KBSSWPG/',
        created_at: '2018-07-06T10:04:23Z',
        hash_sha1: '96239c7247d3dd043d2050dc9c95942089cb1a9f',
        created_by: 11001,
        height: 0,
        width: 0,
        attrs: {
          cdn_url: 'https://f-test.etuschool.org/67BD020K2E6JMGX9R00NZ8DQ8KBSSWPG',
          cdn_uid: 'FpYjnHJH090EPSBQ3JyVlCCJyxqf',
          file_name: '1530871462174.jpg',
          height: 460,
          width: 690,
          duration: '0',
          media_type: 'image/jpeg',
          rotation: 0
        },
        is_original: false,
        id: '67BD020K2E6JMGX9R00NZ8DQ8KBSSWPG',
        size: 53943
      }
    },
    {
      media_id: '73VS29QWEZNRYSVDQX30TV4VFMBBYK37',
      type: 'picture',
      extra: {
        url: '/api/v1/files/73VS29QWEZNRYSVDQX30TV4VFMBBYK37/',
        created_at: '2018-07-06T10:04:23Z',
        hash_sha1: '672f3eb6d34028fb7b9468d0fd1ad87fc154c7ef',
        created_by: 11001,
        height: 0,
        width: 0,
        attrs: {
          cdn_url: 'https://f-test.etuschool.org/73VS29QWEZNRYSVDQX30TV4VFMBBYK37',
          cdn_uid: 'FmcvPrbTQCj7e5Ro0P0a2H_BVMfv',
          file_name: '1530871462178.jpg',
          height: 509,
          width: 720,
          duration: '0',
          media_type: 'image/jpeg',
          rotation: 0
        },
        is_original: false,
        id: '73VS29QWEZNRYSVDQX30TV4VFMBBYK37',
        size: 46835
      }
    }
  ]
  const props = {
    items: medias,
    margin: 5,
    background: '#fff',
    fixed: true,
    noName: true,
    bordered: false,
    lazyLoad: false,
    width: 160,
    height: 120,
    hasDeleteButton: false
  }
  const wrapper = shallow(<MediaList {...props} />)
  const rendered = render(<MediaList {...props} />)
  const renderedSnap = renderer.create(<MediaList {...props} />)
  return {
    medias,
    props,
    rendered,
    renderedSnap,
    wrapper
  }
}

describe('MediaList', function() {
  const { props, medias, renderedSnap, rendered, wrapper } = setup()

  it('correct view', () => {
    expect(rendered.find('.ant-card').length).toEqual(medias.length)
    expect(wrapper.prop('margin')).toEqual(5)

    const style = wrapper
      .dive()
      .childAt(0)
      .prop('style')
    expect(style.width).toEqual(160)
    expect(style.height).toEqual(120)
    expect(
      wrapper
        .dive()
        .childAt(0)
        .prop('bordered')
    ).toEqual(false)
    expect(renderedSnap.toJSON()).toMatchSnapshot()
  })

  it('another view', () => {
    const otherProps = {
      ...props,
      noName: false,
      fixed: false,
      background: '#666',
      width: 40,
      bordered: true,
      lazyLoad: true,
      hasDeleteButton: true
    }
    const v = shallow(<MediaList {...otherProps} />)
    expect(v.find('Icon').length).toEqual(4)
    expect(
      v
        .find('Icon')
        .at(0)
        .prop('type')
    ).toEqual('delete')
    const card = v.childAt(0)
    const cardProps = card.props()

    expect(cardProps).toHaveProperty('background', '#666')
    expect(cardProps).toHaveProperty('fixed', 0)
    expect(cardProps).toHaveProperty('bordered', true)

    const style = card.prop('style')
    expect(style).toHaveProperty('width', 40)
    expect(card.find('FileName').length).toEqual(1)
    expect(v.find('FileIcon').length).toEqual(2)
    expect(v.find('ImageCard').length).toEqual(2)
  })

  it('function', () => {
    expect(MediaList.getFileName({})).toEqual(undefined)
    expect(MediaList.getFileName(medias[0])).toEqual('LastOfUs2.pdf')
    expect(MediaList.getFileName({ extra: { attrs: { file_name: 'world' } } })).toEqual('world')
    const media_id = medias[0].media_id
    expect(MediaList.getFile(media_id)).toEqual(`/bos/api/v1/files/${media_id}/`)
    expect(MediaList.getDownload(media_id)).toEqual(`/bos/api/v1/files/${media_id}/download/`)
    expect(MediaList.getThumbnail(media_id, 100, 50)).toEqual(
      `/bos/api/v1/files/${media_id}/thumbnail/?width=100&height=50&mode=zoom_cut`
    )
  })

  it('FileName', () => {
    const v = shallow(<FileName name={'hello'} fixed />)
    expect(v.prop('fixed')).toEqual(1)
    expect(v.find('Tooltip').length).toEqual(1)
    expect(
      v
        .find('Tooltip')
        .children()
        .text()
    ).toEqual('hello')
  })

  it('List', () => {
    const tree = renderer.create(<List margin={3}><List.Item>hello</List.Item></List>).toJSON()
    expect(tree).toHaveStyleRule('margin', '0 -3px')
    expect(tree.children[0].children[0].children[0]).toEqual('hello')
  })
})
