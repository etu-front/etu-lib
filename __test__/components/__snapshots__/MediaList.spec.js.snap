// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaList correct view 1`] = `
.c0 {
  clear: both;
  margin: 0 -5px;
  list-style: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c0 > div,
.c0 > .ant-card {
  margin: 5px;
  overflow: hidden;
  border-color: #d5d4d4;
  border-radius: 4px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1.ant-card {
  margin: 5px;
  overflow: hidden;
  border-color: #d5d4d4;
  border-radius: 4px;
  background: #fff;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 .ant-card-body {
  padding: 0;
  text-align: center;
  cursor: pointer;
}

.c2 {
  position: absolute;
  left: 0;
  right: 0;
  padding: 0 5px;
  bottom: 0;
  height: 20px;
  font-size: 12px;
  line-height: 20px;
  color: #fff;
  background: #33333399;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  text-align: center;
}

<div
  className="c0"
>
  <div
    background="#fff"
    className="ant-card c1"
    fixed={1}
    margin={5}
    onClick={[Function]}
    style={
      Object {
        "height": 120,
        "width": 160,
      }
    }
  >
    <div
      className="ant-card-body"
      style={Object {}}
    >
      <i
        className="anticon"
        style={
          Object {
            "fontSize": 80,
            "margin": "0 auto",
          }
        }
      >
        <svg
          aria-hidden="true"
          className=""
          fill="currentColor"
          focusable="false"
          height="1em"
          width="1em"
        >
          <use
            xlinkHref="#icon-pdf"
          />
        </svg>
      </i>
      <div
        className="c2"
      >
        <span
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          LastOfUs2.pdf
        </span>
      </div>
    </div>
  </div>
  <div
    background="#fff"
    className="ant-card c1"
    fixed={1}
    margin={5}
    onClick={[Function]}
    style={
      Object {
        "height": 120,
        "width": 160,
      }
    }
  >
    <div
      className="ant-card-body"
      style={Object {}}
    >
      <i
        className="anticon"
        style={
          Object {
            "fontSize": 80,
            "margin": "0 auto",
          }
        }
      >
        <svg
          aria-hidden="true"
          className=""
          fill="currentColor"
          focusable="false"
          height="1em"
          width="1em"
        >
          <use
            xlinkHref="#icon-word"
          />
        </svg>
      </i>
      <div
        className="c2"
      >
        <span
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          【中文版】ETU PBL Backwards Design Template.docx
        </span>
      </div>
    </div>
  </div>
  <div
    background="#fff"
    className="ant-card c1"
    fixed={1}
    margin={5}
    onClick={[Function]}
    style={
      Object {
        "height": 120,
        "width": 160,
      }
    }
  >
    <div
      className="ant-card-body"
      style={Object {}}
    >
      <div
        style={
          Object {
            "backgroundColor": "#F0F0F0",
            "backgroundImage": "url(/bos/api/v1/files/67BD020K2E6JMGX9R00NZ8DQ8KBSSWPG/thumbnail/?width=160&height=120&mode=zoom_cut)",
            "backgroundPosition": "center",
            "backgroundSize": "cover",
            "display": "flex",
            "height": 120,
            "margin": "0 auto",
            "maxWidth": "100%",
            "width": 160,
          }
        }
      />
    </div>
  </div>
  <div
    background="#fff"
    className="ant-card c1"
    fixed={1}
    margin={5}
    onClick={[Function]}
    style={
      Object {
        "height": 120,
        "width": 160,
      }
    }
  >
    <div
      className="ant-card-body"
      style={Object {}}
    >
      <div
        style={
          Object {
            "backgroundColor": "#F0F0F0",
            "backgroundImage": "url(/bos/api/v1/files/73VS29QWEZNRYSVDQX30TV4VFMBBYK37/thumbnail/?width=160&height=120&mode=zoom_cut)",
            "backgroundPosition": "center",
            "backgroundSize": "cover",
            "display": "flex",
            "height": 120,
            "margin": "0 auto",
            "maxWidth": "100%",
            "width": 160,
          }
        }
      />
    </div>
  </div>
</div>
`;
